File: src/main/java/org/hisp/dhis/expression/parse/FragmentContext.java
package org.hisp.dhis.expression.parse;

import java.util.function.Function;

public interface FragmentContext {

        /*
        Lookup named building blocks (non-terminals)
    */

    NonTerminal fragment(String name);

    static NonTerminal lookup(Expr expr, Function<Expr, String> parseName, Function<String, NonTerminal> lookup) {
        int s = expr.position();
        String name = parseName.apply(expr);
        NonTerminal res = lookup.apply(name);
        if (res == null)
        {
            expr.error(s, "name not available in context: "+name);
        }
        return res;
    }
}

File: src/commonMain/kotlin/org/hisp/dhis/lib/expression/syntax/FragmentContext.kt
package org.hisp.dhis.lib.expression.syntax

fun interface FragmentContext {
    /*
        Lookup named building blocks (non-terminals)
    */
    fun fragment(name: String): Fragment?

    companion object {
        fun lookup(expr: Expr, parseName: (Expr) -> String, lookup: (String) -> Fragment?): Fragment {
            val s = expr.position()
            val name = parseName(expr)
            val res = lookup(name)
            if (res == null) {
                expr.error(s, "Unknown function or constant: '$name'")
            }
            return res!!
        }
    }
}
